modular_skyrat/modules/modular_implants/code/nifsofts/station_pass.dm

```diff
@@ -0,0 +1,191 @@
+	///What username is being sent out?
+	var/transmitted_name = ""
+	///What message is being sent to other users?
+	var/transmitted_message = ""
+	///What ckey is being used by the owner? This is mostly here so that messages can't get spammed
+	var/transmitted_identifier = ""
+
+	///What messages has the user recieved?
+	var/list/message_list = list()
+	///The datum that is being used to receive messages
+	var/datum/proximity_monitor/advanced/station_pass/proximity_datum
+
+/datum/nifsoft/station_pass/New()
+	. = ..()
+
+	if(!transmitted_name)
+		transmitted_name = linked_mob.name
+	if(!transmitted_message)
+		transmitted_message = "Hello, I am [transmitted_name], it's nice to meet you!"
+	transmitted_identifier = linked_mob.ckey
+
+	add_message("station_pass_nifsoft", name, "Hello World")
+	proximity_datum = new(linked_mob, 1)
+	proximity_datum.parent_nifsoft = src
+
+/datum/nifsoft/station_pass/New()
+	qdel(proximity_datum)
+	proximity_datum = null
+
+	return ..()
+
+///Adds a message to the message_list based off the name and message
+/datum/nifsoft/station_pass/proc/add_message(sender_identifier, recieved_name, recieved_message)
+	if(!recieved_message)
+		return FALSE
+	if(!name)
+		name = "Unknown User"
+	else
+		for(var/message in message_list)
+			if(message["identifier"] == sender_identifier)
+				message["sender_name"] = recieved_name
+				message["message"] = recieved_message
+				message["timestamp"] = station_time_timestamp()
+				return TRUE
+
+	message_list.Insert(1, list(list(identifier = sender_identifier, sender_name = recieved_name, message = recieved_message, timestamp = station_time_timestamp())))
+	return TRUE
+
+///Removes a message from the message_list based on the message_to_remove
+/datum/nifsoft/station_pass/proc/remove_message(message_to_remove)
+	var/removed_message = locate(message_to_remove) in message_list
+	if(!removed_message)
+		return FALSE
+
+	message_list -= list(removed_message)
+	return TRUE
+
+/datum/nifsoft/station_pass/activate()
+	. = ..()
+	ui_interact(linked_mob)
+
+
+/datum/nifsoft/station_pass/load_persistence_data()
+	. = ..()
+	var/datum/modular_persistence/persistence = .
+	if(!persistence)
+		return FALSE
+
+	transmitted_name = persistence.station_pass_nifsoft_name
+	transmitted_message = persistence.station_pass_nifsoft_message
+
+/datum/nifsoft/station_pass/save_persistence_data(datum/modular_persistence/persistence)
+	. = ..()
+	if(!.)
+		return FALSE
+
+	persistence.station_pass_nifsoft_message = transmitted_message
+	persistence.station_pass_nifsoft_name = transmitted_name
+
+///The proximty_monitor datum used by the station_pass NIFSoft
+/datum/proximity_monitor/advanced/station_pass
+	///What NIFSoft is this currently attached to?
+	var/datum/weakref/parent_nifsoft
+
+/datum/proximity_monitor/advanced/station_pass/on_entered(turf/source, atom/movable/entered)
+	. = ..()
+	if(host == entered)
+		return FALSE
+
+	var/datum/nifsoft/station_pass/recieving_nifsoft = parent_nifsoft
```

Comment:
Weakrefs don't work like this, you should do I think something like `parent_nifsoft.resolve()` or something along those lines to actually get the real thing it's associated to.